### PR TITLE
nuttx: Change all default stack size to DEFAULT_TASK_STACKSIZE

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
yes, the old value is small than 2048, but user can still
change to other value in defconfg if you really want.

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

### CODE REVIEW

File: arch/arm/src/cxd56xx/Kconfig [hunk 1]
```diff
@@ -1117,7 +1117,7 @@ config CXD56_GNSS_NSIGNALRECEIVERS
 
 config CXD56_GNSS_WORKER_STACKSIZE
 	int "GNSS worker thread stack size"
-	default 128
```
Comment (patacongo) [inline arch/arm/src/cxd56xx/Kconfig:1120]: This is an excessive change in stack size.

File: arch/Kconfig [hunk 2]
```diff
@@ -605,7 +605,7 @@ config PAGING_DEFPRIO
 
 config PAGING_STACKSIZE
 	int "Page fill worker thread stack size"
-	default 1024
```
Comment (patacongo) [inline arch/Kconfig:608]: This changes the default

File: arch/arm/src/sam34/Kconfig [hunk 3]
```diff
@@ -1532,7 +1532,7 @@ config WDT_THREAD_PRIORITY
 
 config WDT_THREAD_STACKSIZE
 	int "Watchdog Thread Stacksize"
-	default 1024
```
Comment (patacongo) [inline arch/arm/src/sam34/Kconfig:1535]: This changes the default

File: sched/Kconfig [hunk 4]
```diff
@@ -1766,7 +1766,7 @@ menu "Stack and heap information"
 
 config IDLETHREAD_STACKSIZE
 	int "Idle thread stack size"
-	default 1024
```
Comment (patacongo) [inline sched/Kconfig:1769]: This is not correct.  This stack should not be changed.

File: arch/Kconfig [hunk 5]
```diff
@@ -476,7 +476,7 @@ if ARCH_KERNEL_STACK
 
 config ARCH_KERNEL_STACKSIZE
 	int "Kernel stack size"
-	default 1568
```
Comment (patacongo) [inline arch/Kconfig:479]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/sama5/sama5d2-xult/Kconfig [hunk 6]
```diff
@@ -25,8 +25,7 @@ endchoice # CPU Frequency
 
 config SAMA5D2XULT_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1536 if USBHOST_HUB
-	default 1024 if !USBHOST_HUB
```
Comment (patacongo) [inline boards/arm/sama5/sama5d2-xult/Kconfig:29]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/sama5/sama5d3-xplained/Kconfig [hunk 7]
```diff
@@ -113,8 +113,7 @@ endchoice # AT25 serial FLASH configuration
 
 config SAMA5D3XPLAINED_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1536 if USBHOST_HUB
-	default 1024 if !USBHOST_HUB
```
Comment (patacongo) [inline boards/arm/sama5/sama5d3-xplained/Kconfig:117]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/sama5/sama5d3x-ek/Kconfig [hunk 8]
```diff
@@ -180,8 +180,7 @@ endchoice # AT24 serial EPPROM configuration
 
 config SAMA5D3xEK_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1536 if USBHOST_HUB
-	default 1024 if !USBHOST_HUB
```
Comment (patacongo) [inline boards/arm/sama5/sama5d3x-ek/Kconfig:184]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/sama5/sama5d4-ek/Kconfig [hunk 9]
```diff
@@ -323,8 +323,7 @@ endif # SAMA5D4EK_ROMFS_MOUNT
 
 config SAMA5D4EK_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1536 if USBHOST_HUB
-	default 1024 if !USBHOST_HUB
```
Comment (patacongo) [inline boards/arm/sama5/sama5d4-ek/Kconfig:327]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/stm32/olimex-stm32-p407/Kconfig [hunk 10]
```diff
@@ -7,7 +7,7 @@ if ARCH_BOARD_OLIMEX_STM32P407
 
 config STM32F4DISCO_OLIMEXP407_STACKSIZE
 	int "USB host waiter stack size"
-	default 1024
```
Comment (patacongo) [inline boards/arm/stm32/olimex-stm32-p407/Kconfig:10]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/stm32/photon/Kconfig [hunk 11]
```diff
@@ -46,7 +46,7 @@ config PHOTON_WDG_THREAD_PRIORITY
 
 config PHOTON_WDG_THREAD_STACKSIZE
 	int "Watchdog Thread Stacksize"
-	default 1024
```
Comment (patacongo) [inline boards/arm/stm32/photon/Kconfig:49]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/stm32/stm32f411e-disco/Kconfig [hunk 12]
```diff
@@ -7,7 +7,7 @@ if ARCH_BOARD_STM32F411E_DISCO
 
 config STM32F411DISCO_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1024
```
Comment (patacongo) [inline boards/arm/stm32/stm32f411e-disco/Kconfig:10]: **It is inappropriate to increase the size of this tuned stack size.  Let's not do that.**

File: boards/arm/stm32/stm32f429i-disco/Kconfig [hunk 13]
```diff
@@ -86,7 +86,7 @@ config STM32F429I_DISCO_RAMMTD_SIZE
 
 config STM32F429IDISCO_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1024
```
Comment (patacongo) [inline boards/arm/stm32/stm32f429i-disco/Kconfig:89]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: boards/arm/stm32/stm32f4discovery/Kconfig [hunk 14]
```diff
@@ -36,7 +36,7 @@ config STM32_ROMFS_IMAGEFILE
 
 config STM32F4DISCO_USBHOST_STACKSIZE
 	int "USB host waiter stack size"
-	default 1024
```
Comment (patacongo) [inline boards/arm/stm32/stm32f4discovery/Kconfig:39]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/audio/Kconfig [hunk 15]
```diff
@@ -149,7 +149,7 @@ config CS43L22_NUM_BUFFERS
 
 config CS43L22_WORKER_STACKSIZE
 	int "CS43L22 worker thread stack size"
-	default 768
```
Comment (patacongo) [inline drivers/audio/Kconfig:152]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/audio/Kconfig [hunk 16]
```diff
@@ -79,7 +79,7 @@ config VS1053_DEVICE_COUNT
 
 config VS1053_WORKER_STACKSIZE
 	int "VS1053 worker thread stack size"
-	default 768
```
Comment (patacongo) [inline drivers/audio/Kconfig:82]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/audio/Kconfig [hunk 17]
```diff
@@ -185,7 +185,7 @@ config WM8776_MSG_PRIO
 
 config WM8776_WORKER_STACKSIZE
 	int "WM8776 worker thread stack size"
-	default 768
```
Comment (patacongo) [inline drivers/audio/Kconfig:188]: This is a horrendous increase and very WRONG!
It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/audio/Kconfig [hunk 18]
```diff
@@ -238,7 +238,7 @@ config WM8904_NUM_BUFFERS
 
 config WM8904_WORKER_STACKSIZE
 	int "WM8904 worker thread stack size"
-	default 768
```
Comment (patacongo) [inline drivers/audio/Kconfig:241]: This is a horrendous increase and very WRONG!
It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/net/Kconfig [hunk 19]
```diff
@@ -88,7 +88,7 @@ config TELNET_IOTHREAD_PRIORITY
 
 config TELNET_IOTHREAD_STACKSIZE
 	int "I/O thread stack size"
-	default 1024
```
Comment (patacongo) [inline drivers/net/Kconfig:91]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

File: drivers/usbhost/Kconfig [hunk 20]
```diff
@@ -286,7 +286,7 @@ config HIDKBD_DEFPRIO
 
 config HIDKBD_STACKSIZE
 	int "Polling thread stack size"
-	default 1024
```
Comment (patacongo) [inline drivers/usbhost/Kconfig:289]: It is inappropriate to increase the size of this tuned stack size.  Let's not do that.

**Comment by patacongo** [comment]
After reviewing an trying to comment on all of the problems and risks.  I come to the conclusion that this change is wrong and must not be merged.

**Comment by patacongo** [comment]
I think that the only way that any change like this could be considered would be to update ALL defconfig files effected by the change in default so that they used the old default value explicitly.  That would If you do not do this, such a sweeping change like this would damage manage configurations, especially smaller, memory constrained configurations.

We cannot do that this way.  It is irresponsbile and disrectful to other people.

**Comment by xiaoxiang781216** [comment]
> After reviewing an trying to comment on all of the problems and risks. I come to the conclusion that this change is wrong and must not be merged.

The change make the default value bigger than the old one, so the side effect is the stack memory consumption will a little bigger than before. But the benefit is that user can change all default stack value by CONFIG_DEFAULT_STACKSIZE for the normal task and still can adjust the stack size in the fine tune phase.

**Comment by patacongo** [comment]
There is no real benefit that is a myth.  It bad engineering.  We cannot do this.  It will break many configurations and we should not even be considering it.  It is inconsistent with the "All users matter" inviolable.  Your personal convenience is not as important as the viability of the configurations for all other users.  I will do everything within power to prevent that.

Perhaps you could consider multiple default stack sizes a MEDIUM (1536), SMALL(1024), TINY(512), ULTRA_TINY(256).  Even then, it would be irresponsible to dicard the settings that people have tuned for their platforms.  Any change to default stack sizes MUST reflected in changes to ALL configuration so that they continue to use the tuned stack size.  Or we cannot merge such a change.

This is about the NuttX user community, no your personal convenience.

**Comment by xiaoxiang781216** [comment]
> I think that the only way that any change like this could be considered would be to update ALL defconfig files effected by the change in default so that they used the old default value explicitly. That would If you do not do this, such a sweeping change like this would damage manage configurations, especially smaller, memory constrained configurations.
> 
> We cannot do that this way. It is irresponsbile and disrectful to other people.

My change just want to make sure that we use CONFIG_DEFAULT_STACKSIZE in a consistent way. And assume any config should have some free memory to accomodate this change. I am Ok to not merge this patch.

**Comment by patacongo** [comment]
Some of those small stack sizes, for example, are turned for AVRs that can do everything they need in 128 or 256 bytes of stack and only hafve 8Kb of RAM total.  This kind of change is would be catastrophic for all users of RAM constrained devices.  And that is a significant, large part of the user community.

You need to think more broadly about all users and not just yourself.

**Comment by patacongo** [comment]
> My change just want to make sure that we use CONFIG_DEFAULT_STACKSIZE in a consistent way. And assume any config should have some free memory to accomodate this change. I am Ok to not merge this patch.

Consistent is good bug not as good as technically correct.  Consistent but wrong is bad.  Correct but inconsistent is okay and acceptable.  Correct and consistent is perfect.  This solution is NOT perfect it is technically flawed and, hence, no acceptable.

**Comment by xiaoxiang781216** [comment]
> Some of those small stack sizes, for example, are turned for AVRs that can do everything they need in 128 or 256 bytes of stack and only hafve 8Kb of RAM total. This kind of change is would be catastrophic for all users of RAM constrained devices. And that is a significant, large part of the user community.
> 
> You need to think more broadly about all users and not just yourself.

Actually, the default stack size can save a lot of painful here:
1.For AVR, we can CONFIG_DEFAULT_STACKSIZE to 128 or 256
2.For sim, w can CONFIG_DEFAULT_STACKSIZE to 4096 or 8124
The bad is that it's too late to introduce CONFIG_DEFAULT_STACKSIZE.

**Comment by patacongo** [comment]
Well, there are many things that affect the correct stack size.
1.  The CPU architecture.  CPUs with 32-bit wides stacks need large stack sizes, CPUs with 64-bit wide stacks need larger stack sizes.  CPUs with narrow stack widthes need less stack stize.
2.  Call depth and use of local variables on the stack.  This is not controlable and must be determined on a task-by-task basis.  The Simulator running on x64_86 needs big stacks only because the normal convention is to put large arrays on the stack.  But not all simulator tasks need large stacks.  Only those that call into Linux or Windows.

People have spent a lot of time and effort tuning the default in some cases.  We cannot lose that effort due to some unnecessary generalization.  I will always be opposed to replacing sound engineering with some arbitrary, non-technical conventioni.  That is just bad engineering.

**Comment by patacongo** [comment]
A way of doing this without breaking configurations occurs to me:

1. Run all defconfig files through 'make olddefconfig' to create complete .config files.
2. Save the .config file, probably in the same directory as the board's defconfig file
3. Apply the changes to the stack sizes
4. Then run all of the .config files through 'make savedefconfig' to create new defconfig files that replace the original onces.

This sequence will retain the original stack size used in every configuration and so, should not break anything.

**Comment by xiaoxiang781216** [comment]
Yes, it's workable. I will try this once I find some free time.

**Comment by patacongo** [comment]
> This sequence will retain the original stack size used in every configuration and so, should not break anything.

But if it is your objective to control the stack size via the DEFAULT_TASK_STACKSIZE, then this would not work.  Once all of the non-default settings are placed in the defconfig file, then the default is never used.  So that would not work.

**Pull request closed without merging**